eos-autoupdater/main.c

```diff
@@ -274,6 +316,153 @@ update_stamp_file (guint64 last_successful_update_secs,
+  g_autoptr(GVariant) results_variant = NULL;
+  g_autoptr(GError) error = NULL;
+  guint64 last_changed_usecs = 0;
+  const gchar *update_refspec = NULL;
+  const gchar *update_id = NULL;
+
+  results_path = g_build_filename (state_dir, POLL_RESULTS_NAME, NULL);
+  results_file = g_file_new_for_path (results_path);
+  results_contents = g_file_load_bytes (results_file, NULL, NULL, &error);
+  if (results_contents == NULL)
+    {
+      if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND)) {
+        critical (EOS_UPDATER_STAMP_ERROR_MSGID,
+                  "Failed to read autoupdater poll results file: %s",
+                  error->message);
+      }
+
+      return NULL;
+    }
+
+  results_variant = g_variant_new_from_bytes (G_VARIANT_TYPE ("a{sv}"),
+                                              results_contents,
+                                              FALSE);
+  g_variant_lookup (results_variant, "LastChangedUsecs", "t", &last_changed_usecs);
+  g_variant_lookup (results_variant, "UpdateRefspec", "&s", &update_refspec);
+  g_variant_lookup (results_variant, "UpdateID", "&s", &update_id);
+
+  return poll_results_new (last_changed_usecs, update_refspec, update_id);
+}
+
+/**
+ * write_poll_results_file:
+ * @results: the #PollResults to store
+ *
+ * Writes @results to the stored poll results file. The poll results
+ * file is encoded in #GVariant dictionary format. The #PollResults
+ * fields correspond to the following entries:
+ *
+ *   * #PollResults:last_changed_usecs: `LastChangedUsecs` (`t`)
+ *   * #PollResults:update_refspec: `UpdateRefspec` (`s`)
+ *   * #PollResults:update_id: `UpdateID` (`s`)
+ */
+static void
+write_poll_results_file (PollResults *results)
+{
+  const gchar *state_dir = get_state_dir ();
+  g_autofree gchar *results_path = NULL;
+  g_autoptr(GFile) results_file = NULL;
+  g_auto(GVariantBuilder) builder = G_VARIANT_BUILDER_INIT (G_VARIANT_TYPE ("a{sv}"));
+  g_autoptr(GVariant) results_variant = NULL;
+  g_autoptr(GBytes) results_contents = NULL;
+  g_autoptr(GError) error = NULL;
+
+  g_assert (results != NULL);
+
+  g_variant_builder_add (&builder, "{s@v}", "LastChangedUsecs",
+                         g_variant_new_variant (g_variant_new_uint64 (results->last_changed_usecs)));
+  if (results->update_refspec != NULL)
+    g_variant_builder_add (&builder, "{s@v}", "UpdateRefspec",
+                           g_variant_new_variant (g_variant_new_string (results->update_refspec)));
+  if (results->update_id != NULL)
+    g_variant_builder_add (&builder, "{s@v}", "UpdateID",
+                           g_variant_new_variant (g_variant_new_string (results->update_id)));
+
+  if (mkdir_print_errors (state_dir, 0755) != 0)
+    return;
+
+  results_variant = g_variant_builder_end (&builder);
+  results_contents = g_variant_get_data_as_bytes (results_variant);
+  results_path = g_build_filename (state_dir, POLL_RESULTS_NAME, NULL);
+  results_file = g_file_new_for_path (results_path);
```

Comment:
Since we depend on GLib ≥ 2.56, I think this can be condensed to:
```suggestion
  results_file = g_file_new_build_filename (state_dir, POLL_RESULTS_NAME, NULL);
```

Same in a few other places too

Reply:
I ended up wanting to print the path on errors, so I decided to keep what I had rather than having to free the results of `g_file_get_path` in the error cases.

Reply:
You can use `g_file_peek_path()` instead of `g_file_get_path()` to avoid freeing the results. For local files it returns the internally-used path. For remote files, it builds the path (as with `g_file_get_path()`) but caches it internally.

But since you’ve already reworked this once, don’t feel obliged to rework it again! Just mentioning `g_file_peek_path()` for future reference really.